### PR TITLE
Add arity option for path inputs and outputs

### LINKED ISSUE #969: Singleton list conversion cause quirk behaviour
## Bug report 

Minimal example:

```Nextflow
    input:
    file input_files from featureCounts_to_merge.collect()

    output:
    file 'merged_gene_counts.txt'

    script:
    //if we only have 1 file, just use cat and pipe output to csvtk. Else join all files first, and then remove unwanted column names.
    def single = input_files instanceof Path ? 1 : input_files.size()
    def merge = (single == 1) ? 'cat' : 'csvtk join -t -f "Geneid,Start,Length,End,Chr,Strand,gene_name"'
    """
    $merge $input_files | csvtk cut -t -f "-Start,-Chr,-End,-Length,-Strand" | sed 's/Aligned.sortedByCoord.out.markDups.bam//g' > merged_gene_counts.txt
    """
```

### Expected behavior and actual behavior
I expect `input_files.size()` to return the size of the list, e.g. having `n` elements.

Actually, this depends:

- if its just one element, it returns a Path object, so path.size() returns the length of the path as a String
- if its multiple elements, it returns the list size, as expected

### Steps to reproduce the problem

See above testcase

### Environment 

* Nextflow version: 18.10.1 
* Java version: 1.8
* Operating system: no-arch

### PR DESCRIPTION
**Pull request by bentsherman**
Closes #969 #1236 #2425 #3341

This PR adds an `arity` option for path inputs and outputs. The arity can be a single number (e.g. `1`) or a range (e.g. `1..*`). The arity is used (1) to validate the expected number of files and (2) to determine, when a path contains only one file, whether to "wrap" that path in a list.

Currently, a path with a single file is always unwrapped. With this PR, the path will be unwrapped only if there is a single file *and* the arity is "single", meaning it can have at most 1 element (in other words, `1` and `0..1` are "single").

The default arity is `1`, or `1..*` if the param has a glob pattern. For optional output params, the default then is `0..1` or `0..*` respectively.

This PR introduces a minor breaking change -- a path input/output with a glob pattern will have a default arity of `1..*`, so it will be wrapped in a list even if there is only one file. Without this PR, this file would be unwrapped. I think this breaking change is worthwhile since it is exactly the inconsistency that users were complaining about.

### CODE REVIEW

File: modules/nextflow/src/main/groovy/nextflow/script/params/ArityParam.groovy [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+package nextflow.script.params
+
+
+/**
+ * Implements an arity option for process inputs and outputs.
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+trait ArityParam {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/script/params/ArityParam.groovy:25]: worth adding some unit tests for this trait

**Comment by pditommaso** [comment]
> This PR introduces a minor breaking change -- a path input/output with a glob pattern will have a default arity of 1..*, so it will be wrapped in a list even if there is only one file.

This should be avoided. it could create nasty side effects hard depending the nextflow version. 

Arity should default to `null`, which will implicitly stick to the current behaviour, and use the new rule only when explicitly defined.

**Comment by bentsherman** [comment]
Should be able to make that work. It's too bad though, because the current defaults would probably work 99% of the time without the user having to specify. It would be nice if we could eventually transition to these default settings instead of just `null`.

**Comment by bentsherman** [comment]
ArityParam fails on my laptop but passes in CI 🤷

**Comment by bentsherman** [comment]
I think that is covered by the e2e test I added: [tests/process-arity.nf](https://github.com/nextflow-io/nextflow/pull/3706/files#diff-fd9c080f1d5d455644e7e1da90f3553b1a0508373653974ca6434c707a591ff3)

**Comment by pditommaso** [comment]
Better to add also as unit test

**Comment by pditommaso** [comment]
I found a glitch testing this. If I'm not wrong, when declaring an input `arity: '0..1', it could be a null or a path type, however, I've noticed that when an empty list is given the input type is a list. 

This was my test case 

```
process foo {
  input:
  path x, arity: '0..1'

  script:
  println "inputs [${x.class.simpleName}]=$x"

  """
  echo $x
  """
}


workflow {
  foo([])
}
```

**Comment by bentsherman** [comment]
I think I wasn't sure how to handle that case. You think the list should be coerced into a single null path?

**Comment by bentsherman** [comment]
Now I'm thinking we shouldn't allow the arity to include 0, and instead only support that through the nullable option #2893

**Comment by pditommaso** [comment]
Should be null according this 

> Currently, a path with a single file is always unwrapped. With this PR, the path will be unwrapped only if there is a single file and the arity is "single", meaning it can have at most 1 element (in other words, 1 and 0..1 are "single").

Instead I believe we should drop `nullable` in  favour of this

**Comment by bentsherman** [comment]
> Instead I believe we should drop nullable in favour of this

That could work, actually I think that would be much simpler

> With this PR, the path will be unwrapped only if there is a single file and the arity is "single", meaning it can have at most 1 element (in other words, 1 and 0..1 are "single").

Well, I was only thinking about outputs when I wrote this. If an *input* has `arity = '0..1'`, then it should receive a single file or a "null" file, but not an empty list. An empty list should cause an error.

**Comment by bentsherman** [comment]
Also I think I will add the ability to specify `arity: true` to have the arity inferred from the file pattern. Originally I was doing this by default but this way it can be an opt-in feature.

**Comment by bentsherman** [comment]
I added the `NullPath` from the other PR, and made it fail if a nullable input receives an empty list.

The problem with the null path is that it goes through a bunch of code that it probably shouldn't, and produces this warning:
```
WARN: File system is unable to get file attributes file: nextflow.util.NullPath@7f -- Cause: java.nio.file.ProviderMismatchException
```

**Comment by bentsherman** [comment]
I'm thinking about the [zero one infinity rule](https://en.wikipedia.org/wiki/Zero_one_infinity_rule) and wondering if we should only allow the following:

- `0..1`
- `0..*`
- `1`
- `1..*`

It seems to be that other cases like `1..2` or `2` should be represented with tuples of (possibly nullable) files, for example:

```groovy
// instead of arity: '2'
tuple path('1.txt'), path('2.txt')

// instead of arity: '1..2'
tuple path('1.txt'), path('2.txt', arity: '0..1')
```

**Comment by bentsherman** [comment]
I don't think such a restriction would simplify this PR much, and there might be some weird cases where something like `a..b` is needed, so maybe better to only suggest it as a best practice.

**Comment by pditommaso** [comment]
I've merged this manually, removing the support for `0..1` that opens the problem of how to handle the optional path. It can be added in the following iteration. See 42504d3c.

**Comment by bentsherman** [comment]
Thanks. I will rebase this branch and make a new PR with just the optional features

**Pull request closed without merging**